aiohttp/cookiejar.py

```diff
@@ -261,8 +262,34 @@ def filter_cookies(self, request_url: URL = URL()) -> "BaseCookie[str]":
                 request_origin = request_url.origin()
             is_not_secure = request_origin not in self._treat_as_secure_origin
 
+        # d: domain, d could be subdomain
+        # p: path
+        d = hostname
+
+        pairs = []
+        while d:
+            p = request_url.path
+            while p:
+                pairs.append((d, p))
+                p = p.rsplit("/", maxsplit=1)[0]
+
+            try:
+                d = d.split(".", maxsplit=1)[1]
+            except IndexError:
+                # handle last element for split
+                d = ""
+
+        # shared cookie, it should have max of 1 entry
+        pairs.append(("", "/"))
+
+        cookies = itertools.chain.from_iterable(
+            self._cookies[p].values() for p in pairs
```

Comment:
Looking at the tests, this needs to be reversed, so longer paths are done later and therefore overwrite shorter paths.
```suggestion
            self._cookies[p].values() for p in reversed(pairs)
```